Live: expose HTTP push endpoint that will read influx line protocol and publish to websocket

FZambia opened the pull request:
**What this PR does / why we need it**:

If Live feature on we can start receiving metrics from Telegraf configured like this:

```toml
[agent]
  interval = "1s"
  flush_interval = "1s"

[[inputs.system]]
  fielddrop = ["uptime_format", "uptime"]
  [inputs.system.tags]
    mylabel = "boom"

[[inputs.cpu]]
  totalcpu = true

[[outputs.http]]
  url = "http://localhost:3000/api/live/push/mypath"
  method = "POST"
  data_format = "influx"
  content_encoding = "identity"

  [outputs.http.headers]
    Authorization = "Bearer <Grafana API Token>"
```

Input will be converted to `data.Frame`s and published to channels.


To see the data in the browser, use the `-- Grafana --` datasource with "Live Measurements" and enter the channel path.  The path will relate to your request and measurement name

![image](https://user-images.githubusercontent.com/705951/113543852-14b86e00-959c-11eb-892f-8d1890941586.png)

FZambia committed:
start work on streaming plugin support

FZambia committed:
cleanups and comments

FZambia committed:
move code to separate files, add constructors

FZambia committed:
do not use method for getting legacy data plugin

FZambia committed:
revert changes in GetDataPlugin

FZambia committed:
stream manager entity

FZambia committed:
minor code cleanups

FZambia committed:
fix build

FZambia committed:
start prototyping

FZambia committed:
Merge branch 'master' into FZambia/streaming_plugin_merge

FZambia committed:
migrate to data frame

FZambia committed:
move testdata to plugin live scope

FZambia committed:
migrate to datasource uid

FZambia committed:
fix tests

FZambia committed:
lint fixes

FZambia committed:
more test fixes

FZambia committed:
Merge branch 'master' into FZambia/streaming_plugin

FZambia committed:
20Hz test data stream

FZambia committed:
Merge branch 'FZambia/streaming_plugin' into FZambia/streaming_telegraf

FZambia committed:
code cleanups

FZambia committed:
experiment with format

FZambia committed:
Update pkg/services/live/channel.go

Co-authored-by: Ryan McKinley <ryantxu@gmail.com>

FZambia committed:
refactor telegraf conversion

FZambia committed:
tests, comments, review fixes

FZambia committed:
Merge branch 'FZambia/streaming_plugin' of github.com:grafana/grafana into FZambia/streaming_plugin

FZambia committed:
Merge branch 'master' into FZambia/streaming_plugin

FZambia committed:
Merge branch 'FZambia/streaming_plugin' into FZambia/streaming_telegraf

FZambia committed:
Merge remote-tracking branch 'origin' into FZambia/streaming_plugin

FZambia committed:
up plugin sdk dependency

FZambia committed:
renamings

FZambia committed:
Merge branch 'FZambia/streaming_plugin' into FZambia/streaming_telegraf

FZambia committed:
pin to live-sdk v0.0.1

FZambia committed:
Merge branch 'master' into FZambia/streaming_telegraf

FZambia committed:
protect with auth

FZambia committed:
cleanups, renamings

FZambia committed:
Merge branch 'master' into FZambia/streaming_telegraf

FZambia committed:
update to use labels column format

FZambia committed:
invert format: use wide by default

FZambia commented on pkg/api/api.go:
```diff
@@ -399,6 +399,10 @@ func (hs *HTTPServer) registerRoutes() {
 		apiRoute.Post("/short-urls", bind(dtos.CreateShortURLCmd{}), routing.Wrap(hs.createShortURL))
+
+		if hs.Live.IsEnabled() {
+			apiRoute.Post("/live/telemetry/*", hs.LiveTelemetry.Handle)
```
not very happy to pass channel path as part of API url here - looks like alternatives are custom header or url param.

ryantxu replied:
do we want "telemetry" as an additional scope?  if so, does this path conflict?

FZambia replied:
At this moment everything published to this endpoint goes to `telemetry/{PATH}/{METRIC_NAME}` where `{PATH}` is `*` from route handler path `/live/telemetry/*`.

So:
* POST to `/live/telemetry/test` will result into publish into `telemetry/test/cpu`, `telemetry/test/system` 
* POST to `/live/telemetry/test/2` will result into publish into `telemetry/test/2/cpu`, `telemetry/test/2/system`

But it's not obvious to me whether this way of channel construction is ok or not.

ryantxu replied:
I thin we should require that the POST 'path' component does not contain a slash -- ie `POST to /live/telemetry/test/2` should be an error, but POST to /live/telemetry/test-2 is OK

ryantxu replied:
maybe:
```
                if hs.Live.IsEnabled() {
			apiRoute.Post("/live/push/:id", hs.LiveTelemetry.Handle)
		}
		```

where id has to be a simple string?

FZambia committed:
Merge branch 'master' into FZambia/streaming_telegraf

FZambia committed:
up live sdk

tonychoe committed:
Configure the Jaeger data source with provisioning (#32620)

* Configure the Jaeger data source with provisioning

We should have a good instruction and example for configuring the Jaeger data source with provisioning. It was hard to find how to add the tarce-to-logs, so I made one.

* Update docs/sources/datasources/jaeger.md

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

kylebrandt committed:
ngalerting: represent nil/empty labels the same (#32652)

was getting duplicates of [] and null before

FZambia committed:
Merge branch 'master' into FZambia/streaming_telegraf

FZambia committed:
add schema caching

FZambia committed:
send schema if not exists in cache yet

FZambia committed:
do not return error is schema not exists

ryantxu committed:
add push scope

ryantxu committed:
add frontend scope

FZambia committed:
tmp fix for internal error

FZambia committed:
properly pass plugin name

ryantxu committed:
2nd level cache

ryantxu committed:
2nd level cache

ryantxu committed:
Reducers: minor reducer cleanup (#32656)

ryantxu committed:
Live: throttle messages when FPS decreases (#32627)

* throttle when FPS is low

* fix throttling

* grafanaStreamingPerfBudget

* grafanaStreamingPerfBudget

* change global strategy

* also throttle frontend

Co-authored-by: Leon Sorokin <leeoniya@gmail.com>

ryantxu committed:
use managed streams

ryantxu committed:
use managed streams

ryantxu committed:
send data on connection

ryantxu committed:
show field selector

ryantxu committed:
query buffer

ryantxu committed:
don't update with same structure

ryantxu committed:
avoid re-render when the data is the same

ryantxu committed:
Merge remote-tracking branch 'origin/master' into FZambia/streaming_telegraf

ryantxu committed:
lint fix

ryantxu committed:
fix spelling

ryantxu committed:
more explicit subscriptions

ryantxu committed:
merge master

ryantxu committed:
less logging

FZambia committed:
various cleanups

ryantxu reviewed (approved):
Looks great!
